openid-4-verifiable-presentations-1_0.md

```diff
@@ -252,10 +252,10 @@ This specification enables the Verifier to send both Presentation Definition JSO
 This specification defines the following new parameters:
 
 `presentation_definition`:
-: A string containing a Presentation Definition JSON object. See (#request_presentation_definition) for more details. This parameter MUST be present when `presentation_definition_uri` parameter, or a `scope` value representing a Presentation Definition is not present.
+: OPTIONAL. A string containing a Presentation Definition JSON object. See (#request_presentation_definition) for more details. This parameter MUST NOT be present when `presentation_definition_uri` parameter, or a `scope` value representing a Presentation Definition is present.
 
 `presentation_definition_uri`:
-: A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. This parameter MUST be present when `presentation_definition` parameter, or a `scope` value representing a Presentation Definition is not present. See (#request_presentation_definition_uri) for more details.
+: OPTIONAL. A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. This parameter MUST NOT be present when `presentation_definition` parameter, or a `scope` value representing a Presentation Definition is present. See (#request_presentation_definition_uri) for more details.
 
```

Comment:
```suggestion

One of the following parameters MUST be present to indicate what credential is being requested: a `presentation_definition` parameter, `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter defined by this specification for that purpose.

```

Reply:
I think the exact text proposed would prohibit us from introducing a new query language in a future specification without being backwards incompatible. 
The current specification draft does not contain "any other parameter by this specification for that purpose". If this text would be the 1.0 version it means that in order to be compliant with 1.0 you have to use one of the three mechanisms described. 
So even if a future version of the mechanism does describe a new parameter, like vp_query, actually using that would mean you would not be compliant with this proposed language. And therefore using vp_query would not be backwards compatible with 1.0.

It is important that there is always a request parameter present, so we could consider adding the following:
"One of the following parameters MUST be present to indicate what credential is being requested: a `presentation_definition` parameter, `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter for that purpose."

Reply:
`any other parameter defined by this specification for that purpose.` is what intends to allow us to introduce a new query language in a future specification without breaking changes.

Reply:
@Sakurann The problem that @martijnharing has identified, and that I agree with, is the wording "by this specification". This means v1.0 of the specification, and only v1.0 when it is the only published specification. Thus implementors will code for 3 alternative parameters only, since v1.0 only defines 3 parameters. When a later specification comes along that specifies a 4th parameter, v1.0 implementations will be incompatible with this and will say "protocol error" when the 4th parameter is encountered (or may crash!), since from its perspective there are only 3 standardised values. What @martijnharing's proposed edit does is ensures that v1.0 implementations will also code up for any unknown parameter (in v1.0) that might be defined in the future and so should gracefully handle this by replying "<future value> query not supported".

Reply:
David clarified the issue. I'll try to be more precise:
The issue with the proposed text comes only from it not allowing us to add vp_query in 1.1. If that's not the case then there is no issue, but the reasoning is:

Assumption 1: A change is considered a breaking change if it allows an implementation to be compliant with 1.1 while that implementation is not compliant with 1.0
Assumption 2: It's not allowed to make changes to v1.1 that are considered a breaking change.

Since "this specification" means "OpenID4VP v1.0" in the context of v1.0, the sentence "or any other parameter defined by this specification for that purpose" doesn't add any options because the 3 options listed in that paragraph are the only options defined in OpenID4VP v1.0. Therefore an implementation of a RP according to v1.0 must use one of these 3 options.

If we add the vp_query element to v1.1, then for an RP implementation, using vp_query instead of one of the other 3 options is allowed, since "or any other parameter defined by this specification for that purpose" translates to "or vp_query" in the context of v1.1. 

In this example, this means that: 
- v1.0 effectively says that an RP must use one of 3 options and using anything else is not allowed.
- v1.1 effectively says that an RP must use one of 4 options and using anything else in not allowed.

Since an implementation of an RP under v1.1 using the fourth option (vp_query) is compliant with v1.1 but not compliant with v1.0, actually adding vp_query to v1.1 would be considered a breaking change under assumption 1. 
Therefore under assumption 2 adding vp_query to v1.1 would not be allowed.

If assumption 1 or assumption 2 is not correct (and I happily stand corrected) the concern could very well not apply.

Reply:
Discussed on today's call, change to "defined by this specification (or any future versions of this specification)"

Reply:
@martijnharing could you please accept?

Reply:
Added the language as was discussed in the call to the PR. I do think it is awkward that compliance to this specification now becomes dependent on the publication schedule of future documents. e.g. test tooling for 1.0 can never be finalized since it is always possible that it must be updated when a new version of this specification becomes available.